src/android/CHIPTool/app/src/main/java/com/google/chip/chiptool/commissioner/CommissioningFragment.java

```diff
@@ -0,0 +1,169 @@
+
+    view.findViewById(R.id.cancel_button)
+        .setOnClickListener(
+            new View.OnClickListener() {
+              @Override
+              public void onClick(View view) {
+                WorkManager.getInstance(getActivity())
+                    .cancelWorkById(commssionerWorkRequest.getId());
+
+                CommissionerActivity commissionerActivity = (CommissionerActivity) getActivity();
+                commissionerActivity.finishCommissioning(Activity.RESULT_CANCELED);
+              }
+            });
+
+    view.findViewById(R.id.done_button)
+        .setOnClickListener(
+            new View.OnClickListener() {
+              @Override
+              public void onClick(View view) {
+                CommissionerActivity commissionerActivity = (CommissionerActivity) getActivity();
+                commissionerActivity.finishCommissioning(Activity.RESULT_OK);
+              }
+            });
+  }
+
+  @Override
+  public void onChanged(@Nullable WorkInfo workInfo) {
+    TextView statusText = getActivity().findViewById(R.id.status_text);
+    ProgressBar progressBar = getActivity().findViewById(R.id.commissioning_progress);
+
+    if (workInfo != null) {
+      if (workInfo.getState().isFinished()) {
+
+        showCommissionDone(
+            workInfo.getOutputData().getBoolean(Constants.KEY_SUCCESS, false),
+            workInfo.getOutputData().getString(Constants.KEY_COMMISSIONING_STATUS));
+      } else {
+        showInProgress(workInfo.getProgress().getString(Constants.KEY_COMMISSIONING_STATUS));
+      }
+    }
+  }
+
+  private void showInProgress(String status) {
+    TextView statusText = getActivity().findViewById(R.id.status_text);
+    if (status != null) {
+      statusText.setText(status);
+    }
+
+    ProgressBar progressBar = getActivity().findViewById(R.id.commissioning_progress);
+    progressBar.setVisibility(View.VISIBLE);
+
+    Button cancelButton = getActivity().findViewById(R.id.cancel_button);
+    cancelButton.setVisibility(View.VISIBLE);
+
+    ImageView doneImage = getActivity().findViewById(R.id.done_image);
+    doneImage.setVisibility(View.GONE);
+
+    ImageView errorImage = getActivity().findViewById(R.id.error_image);
+    errorImage.setVisibility(View.GONE);
+
+    Button doneButton = getActivity().findViewById(R.id.done_button);
+    doneButton.setVisibility(View.GONE);
+  }
+
+  private void showCommissionDone(Boolean success, String status) {
+    TextView statusText = getActivity().findViewById(R.id.status_text);
+    if (status != null) {
+      statusText.setText(status);
+    }
+
+    ProgressBar progressBar = getActivity().findViewById(R.id.commissioning_progress);
+    progressBar.setVisibility(View.GONE);
+
+    Button cancelButton = getActivity().findViewById(R.id.cancel_button);
+    cancelButton.setVisibility(View.GONE);
+
+    if (success) {
+      ImageView doneImage = getActivity().findViewById(R.id.done_image);
+      doneImage.setVisibility(View.VISIBLE);
```

Comment:
It would be good to set the visibility of "errorImage" to GONE in the success case and that of "doneImage" set to GONE in the failure case.

Reply:
updated.